src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -244,36 +266,90 @@ QueryPlanPtr MergeTreeDataSelectExecutor::readFromParts(
 
     /// Select the parts in which there can be data that satisfy `minmax_idx_condition` and that match the condition on `_part`,
     ///  as well as `max_block_number_to_read`.
+    /// Skip parts uuids if any to the query context, or skip parts which uuids marked as excluded.
     {
-        auto prev_parts = parts;
-        parts.clear();
+        Context & query_context = context.hasQueryContext() ?
+                                    const_cast<Context &>(context).getQueryContext() : const_cast<Context &>(context);
 
-        for (const auto & part : prev_parts)
-        {
-            if (part_values.find(part->name) == part_values.end())
-                continue;
+        auto process_parts = [&]() -> bool {
+            auto ignored_part_uuids = query_context.getIgnoredPartUuids();
+            std::set<UUID> temp_part_uuids;
 
-            if (part->isEmpty())
-                continue;
+            auto prev_parts = parts;
+            parts.clear();
 
-            if (minmax_idx_condition && !minmax_idx_condition->checkInHyperrectangle(
-                    part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
-                continue;
-
-            if (partition_pruner)
+            for (const auto & part : prev_parts)
             {
-                if (partition_pruner->canBePruned(part))
+                if (part_values.find(part->name) == part_values.end())
+                    continue;
+
+                if (part->isEmpty())
+                    continue;
+
+                if (minmax_idx_condition
+                    && !minmax_idx_condition->checkInHyperrectangle(part->minmax_idx.hyperrectangle, data.minmax_idx_column_types)
+                            .can_be_true)
                     continue;
+
+                if (partition_pruner)
+                {
+                    if (partition_pruner->canBePruned(part))
+                        continue;
+                }
+
+                if (max_block_numbers_to_read)
+                {
+                    auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
+                    if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                        continue;
+                }
+
+                /// populate UUIDs and exclude ignored parts if enabled
+                if (query_context.getSettingsRef().allow_experimental_query_deduplication && part->uuid != UUIDHelpers::Nil)
```

Comment:
As far as I know, currently, all parts have UUIDs. So we will send all of them each time?